File: 2024/day04_ceres_search/main.py
import re
import argparse
from collections import defaultdict


def get_diagonals_and_anti_diagonals(matrix):
    diagonals = defaultdict(list)
    anti_diagonals = defaultdict(list)
    for r, row in enumerate(matrix):
        for c, val in enumerate(row):
            diagonals[r - c].append(val)
            anti_diagonals[r + c].append(val)
    return list(diagonals.values()), list(anti_diagonals.values())


def read_input(filename):
    with open(filename, 'r') as file:
        input = [list(line.strip()) for line in file]

    rows = " ".join(["".join(r) for r in input])
    cols = " ".join(["".join(line) for line in [list(col)
                    for col in list(zip(*input))]])
    diag, anti_diag = get_diagonals_and_anti_diagonals(input)
    diag = " ".join(["".join(line) for line in diag])
    anti_diag = " ".join(["".join(line) for line in anti_diag])

    return input, rows, cols, diag, anti_diag


def part1(rows, cols, diag, anti_diag):
    total_input = " ".join([rows, cols, diag, anti_diag])
    xmas_count = len(re.findall("XMAS", total_input))
    xmas_count += len(re.findall("SAMX", total_input))
    return xmas_count


def part2(input):
    mas_count = 0
    for i in range(1, len(input)-1):
        for j in range(1, len(input[0])-1):
            if input[i][j] != 'A':
                continue
            diag = "".join([input[i-1][j-1], 'A', input[i+1][j+1]])
            anti_diag = "".join([input[i+1][j-1], 'A', input[i-1][j+1]])
            if (diag == 'MAS' or diag == 'SAM') and (anti_diag == 'MAS' or anti_diag == 'SAM'):
                mas_count += 1
    return mas_count


if __name__ == '__main__':
    parser = argparse.ArgumentParser("main.py")
    parser.add_argument("filename", nargs='?',
                        help="input filename", default="input.txt")
    args = parser.parse_args()

    input, rows, cols, diag, anti_diag = read_input(args.filename)
    xmas_count = part1(rows, cols, diag, anti_diag)
    mas_count = part2(input)

    print(f"Part1: xmas_count = {xmas_count}")
    print(f"Part2: mas_count = {mas_count}")

File: 2024/day04_ceres_search/main.rs
use std::{env, fs};

fn read_input(filename: &str) -> Vec<Vec<char>> {
    fs::read_to_string(filename).expect("Unable to open file")
        .lines()
        .map(|line| line.chars().collect())
        .collect()
}

const DIRECTIONS: &[(i32, i32)] = &[
    (-1, -1), (0, -1), (1, -1),
    (-1,  0),          (1,  0),
    (-1,  1), (0,  1), (1,  1),
];

fn part1(input: &Vec<Vec<char>>) -> u32 {
    let mut xmas_ct = 0u32;
    for (r, row) in input.iter().enumerate() {
        for (c, &cell) in row.iter().enumerate() {
            if cell == 'X' {
                for d in DIRECTIONS {
                    let mut x = r as i32;
                    let mut y = c as i32;
                    let mut s = String::new();
                    for _ in 0..4 {
                        if x < 0 || y < 0 || x >= input.len() as i32 || y >= row.len() as i32 {
                            break;
                        }
                        s.push(input[x as usize][y as usize]);
                        x += d.0;
                        y += d.1;
                    }
                    if s == "XMAS" {
                        xmas_ct += 1;
                    }
                }
            }
        }
    }
    xmas_ct
}

fn part2(input: &Vec<Vec<char>>) -> u32 {
    let mut mas_ct = 0;
    for r in 1..input.len()-1 {
        for c in 1..input[0].len()-1 {
            if input[r][c] == 'A' {
                let nw = input[r-1][c-1];
                let se = input[r+1][c+1];
                let ne = input[r+1][c-1];
                let sw = input[r-1][c+1];
                let diag = format!("{nw}{se}");
                let anti_diag = format!("{ne}{sw}");
                if (diag == "MS" || diag == "SM") && (anti_diag == "MS" || anti_diag == "SM") {
                    mas_ct += 1;
                }
            }
        }
    }
    mas_ct
}

fn main() {
    let mut filename = "input.txt";
    let args: Vec<String> = env::args().collect();
    if args.len() == 2 {
        filename = &args[1];
    }

    let input = read_input(filename);
    let xmas_ct = part1(&input);
    let mas_ct = part2(&input);

    println!("Part1: xmas_count = {}", xmas_ct);
    println!("Part1: mas_count = {}", mas_ct);
}

